### PR TITLE
Extension Typing shim 

### PR DESCRIPTION
**Pull request by DedSec256**
We'd like to be able to instantiate and host type providers out-of-process. 
This will allow us to isolate the work of the IDE from the work of type providers and use runtime that is different to the one that hosts FCS.
This PR adds an extension typing shim that allows to override FCS type providers instantiating for the aforementioned purpose

### CODE REVIEW

**Comment by auduchinok** [comment]
There's a failing test `Test Project25 symbol uses of type-provided members` that looks related:
https://dev.azure.com/dnceng/public/_build/results?buildId=486778&view=logs&j=c3b5cd2e-ddad-540a-7ae1-df25e52fdba0&t=d4985d65-7d01-5e76-2879-f1b5d6535baa

**Comment by DedSec256** [comment]
> There's a failing test `Test Project25 symbol uses of type-provided members` that looks related:
> https://dev.azure.com/dnceng/public/_build/results?buildId=486778&view=logs&j=c3b5cd2e-ddad-540a-7ae1-df25e52fdba0&t=d4985d65-7d01-5e76-2879-f1b5d6535baa

image

Locally these tests are green.

**Comment by cartermp** [comment]
Unfortunately we have some heisenbugs.

**Comment by DedSec256** [comment]
> Unfortunately we have some heisenbugs.

@cartermp, detected & fixed 😎 (thanks to @auduchinok)

**Comment by dsyme** [comment]
Could you post an example of what out-of-proc hosting looks like?


This is still in-proc hosting, if I understand correctly? https://github.com/dotnet/fsharp/pull/8235/files#diff-f6b587f1e431a5977102d2a20c2a1990R29

**Comment by DedSec256** [comment]
> Could you post an example of what out-of-proc hosting looks like?
> 
> This is still in-proc hosting, if I understand correctly? https://github.com/dotnet/fsharp/pull/8235/files#diff-f6b587f1e431a5977102d2a20c2a1990R29

@dsyme, the shim is a part of out-of-process type providers protocol model which can be divided in two parts: the in-process one and the out-of-process one. ``InstantiateTypeProvidersOfAssembly(...)`` is called in the in-process area and in the current implementation it creates another process that uses the desired user .NET runtime, where real type providers entities are be created and live. It then create provided types proxies that live in the in-process area and replace the real types. These in-process provided types get info needed for FCS via custom protocol model.

image

**Comment by DedSec256** [comment]
https://github.com/dotnet/fsharp/pull/8235

**Comment by DedSec256** [comment]
- Added more necessary extension points
- Encapsulated work with Shim inside ExtensionTyping
- Removed Tainted as result type from InstantiateTypeProvidersOfAssembly

**Comment by DedSec256** [comment]
If you don't want to make Tainted API public, I can try refactoring the TaintedEquals/GetHashCode in such a way as to remove the Tainted <_> from the arguments and split ExntensionTyping module up into public (Shim and ProvidedModels only) part & internal one

**Comment by dsyme** [comment]
> @dsyme, the shim is a part of out-of-process type providers protocol model which can be divided in two parts: the in-process one and the out-of-process one. 

So I'd like to see a bit more evidence that the proxying model works for this proposed extension point to the compiler.

Could you supply a dummy implementation of proxying as a test case as part of this PR?  

We kind of need evidence that the extension point fills the need it's designed to fill for out-of-proc cases

**Comment by DedSec256** [comment]
> Could you supply a dummy implementation of proxying as a test case as part of this PR?

Ok, I will prepare it and answer!

File: tests/service/ExtensionTypingProviderTests.fs [hunk 1]
```diff
@@ -15,13 +18,65 @@ open FSharp.Compiler.AbstractIL.Internal.Library
 open FSharp.Compiler.Service.Tests.Common
 open Microsoft.FSharp.Core.CompilerServices
 
+module internal XmlProxyProvidersProject = 
+    let fileName1 = Path.ChangeExtension(Path.GetTempFileName(), ".fs")
```
Comment (auduchinok) [inline tests/service/ExtensionTypingProviderTests.fs:22]: Could it use something like the following?
https://github.com/dotnet/fsharp/blob/ad112ad140717481d86e460fe8a646b5c4c17c47/tests/service/Common.fs#L170

**Comment by DedSec256** [comment]
@dsyme, I added a simple test demonstrating how extension points work within a single process. To transfer data about the provided models between processes, you must additionally make the provided models virtual to create your own proxy-provided-models implementations.

**Comment by cartermp** [comment]
@dsyme could you take another look?

**Comment by cartermp** [comment]
@dsyme could you take a look at this?

**Comment by DedSec256** [comment]
@cartermp, I will push the latest changes related to this PR shortly and open another PR closely related to the virtualization of provided entities (:

**Comment by KevinRansom** [comment]
I'm closing this it looks like it's dead.  Please feel free to reopen with a comment, if you wish to pursue it.

Kevin

**Comment by auduchinok** [comment]
@KevinRansom It still awaits a review by @dsyme AFAICT.

**Comment by dsyme** [comment]
I will leave some review comments

However I really need to have a discussion with @vzarytovskii , @KevinRansom and @TIHan about whether this is the hook we want. Notes below

- The PR adds a global hook into FCS to intercept all type provider instantiations, with the intention to use this to run the type providers out of process.  
- The PR has testing, but doesn't actually demonstrate that this allows TPs to really, truly be run out of process

On the whole I'm sympathetic to this.  Type providers are an extension point that can cause arbitrary instability and faults and it is reasonable to have a general mechanism to run them outside the main checking process (even if that itself is outside the IDE) without requiring the modification of each type provider.  It is a real problem that needs solving.

However it feels like we really need evidence that this actually works in practice, and that the architecture doesn't need to be changed further to enable effective out-of-proc.

**Comment by auduchinok** [comment]
> However it feels like we really need evidence that this actually works in practice

@dsyme Due the runtime that Rider uses to host FCS is often different to the runtime used by the compiler invoked by MSBuild, there were many issues about various type providers not working in the IDE. A combination of this PR and https://github.com/dotnet/fsharp/pull/11110 has been working in Rider since the previous release and the most of the issues we've encountered before were fixed with this shim. Here're examples of the issues fixed: [RIDER-31245](https://youtrack.jetbrains.com/issue/RIDER-31245), [RIDER-38184](https://youtrack.jetbrains.com/issue/RIDER-38184), [RIDER-54377](https://youtrack.jetbrains.com/issue/RIDER-54377).

In addition to verifying the fix by hand, there're tests covering some popular and/or interesting for testing providers too: https://github.com/JetBrains/resharper-fsharp/tree/net212/rider-fsharp/src/test/kotlin/typeProviders

> the architecture doesn't need to be changed further to enable effective out-of-proc.

We'll be happy to help evolving it in future if needed. Having that said, having it merged upstream will simplify maintaining a lot, since if there're some changes that break this approach, we'll have to come here and create an issue and work on a fix, likely involving the F# team in some way, anyway.

File: src/fsharp/ExtensionTyping.fs [hunk 2]
```diff
@@ -751,9 +745,108 @@ module internal ExtensionTyping =
         override _.Equals y = match y with :? ProvidedVar as y -> x.Equals y.Handle | _ -> false
         override _.GetHashCode() = x.GetHashCode()
 
+    [<AutoOpen>]
+    module Shim =
+
+        type IExtensionTypingProvider =
+            abstract InstantiateTypeProvidersOfAssembly:
```
Comment (dsyme) [inline src/fsharp/ExtensionTyping.fs:752]: Please put all these arguments into a single object called `ExtensionTypingProviderContext`

File: src/fsharp/ExtensionTyping.fs [hunk 3]
```diff
@@ -751,9 +745,108 @@ module internal ExtensionTyping =
...
+                    pn.ResolveTypeName typeName |> ProvidedType.CreateNoContext
+
+                member this.GetInvokerExpression(provider: ITypeProvider, methodBase: ProvidedMethodBase, paramExprs: ProvidedVar[]) =
+                    provider.GetInvokerExpression(methodBase.Handle, [| for p in paramExprs -> Quotations.Expr.Var (p.Handle) |]) |> ProvidedExpr.Create methodBase.Context
+
+                member this.DisplayNameOfTypeProvider(tp: ITypeProvider, fullName: bool) =
+                    if fullName then tp.GetType().FullName else tp.GetType().Name
+
+        let mutable ExtensionTypingProvider = DefaultExtensionTypingProvider() :> IExtensionTypingProvider
+        let shimLogger (tpe: TypeProviderError) =
```
Comment (dsyme) [inline src/fsharp/ExtensionTyping.fs:809]: This should either be internal or named more apprioriately

File: src/fsharp/ExtensionTyping.fsi [hunk 4]
```diff
@@ -12,7 +12,7 @@ open FSharp.Core.CompilerServices
 open FSharp.Compiler.AbstractIL.IL
 open FSharp.Compiler.Text
 
-module internal ExtensionTyping =
```
Comment (dsyme) [inline src/fsharp/ExtensionTyping.fsi:15]: This should not be public.  Please update the baselines for FCS surface area so we could see properly what's been added

File: src/fsharp/tainted.fs [hunk 5]
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
             for msg in errors do
                 f (new TypeProviderError(errNum, tpDesignation, m, [msg], typeNameContext, methodNameContext))
 
-type TaintedContext = { TypeProvider: ITypeProvider; TypeProviderAssemblyRef: ILScopeRef; Lock: TypeProviderLock }
+type TaintedContext = { TypeProvider : ITypeProvider; TypeProviderAssemblyRef : ILScopeRef; TypeProviderDesignation: string; Lock : TypeProviderLock }
 
-[<NoEquality>][<NoComparison>] 
-type internal Tainted<'T> (context: TaintedContext, value: 'T) =
```
Comment (dsyme) [inline src/fsharp/tainted.fs:83]: I don't believe we want `Tainted` to be public

File: src/fsharp/ExtensionTyping.fsi [hunk 6]
```diff
@@ -337,4 +340,33 @@ module internal ExtensionTyping =
     /// We check by seeing if the type is absent from the remapping context.
     val IsGeneratedTypeDirectReference         : Tainted<ProvidedType> * range -> bool
 
+    [<AutoOpen>]
+    module Shim =
+
+        type IExtensionTypingProvider =
```
Comment (dsyme) [inline src/fsharp/ExtensionTyping.fsi:346]: Please add `///` comments to everything made public

File: src/fsharp/ExtensionTyping.fsi [hunk 7]
```diff
@@ -337,4 +340,33 @@ module internal ExtensionTyping =
...
+
+             /// Find and instantiate the set of ITypeProvider components for the given assembly reference
+            abstract InstantiateTypeProvidersOfAssembly :
```
Comment (dsyme) [inline src/fsharp/ExtensionTyping.fsi:349]: Please give names and `///` comments to all parameters in a public API

File: src/fsharp/tainted.fs [hunk 8]
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
             for msg in errors do
                 f (new TypeProviderError(errNum, tpDesignation, m, [msg], typeNameContext, methodNameContext))
 
-type TaintedContext = { TypeProvider: ITypeProvider; TypeProviderAssemblyRef: ILScopeRef; Lock: TypeProviderLock }
+type TaintedContext = { TypeProvider : ITypeProvider; TypeProviderAssemblyRef : ILScopeRef; TypeProviderDesignation: string; Lock : TypeProviderLock }
```
Comment (dsyme) [inline src/fsharp/tainted.fs:80]: No space before `:` please

File: src/fsharp/tainted.fs [hunk 9]
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
...
 
-[<NoEquality>][<NoComparison>] 
-type internal Tainted<'T> (context: TaintedContext, value: 'T) =
+[<NoEquality>][<NoComparison>]
+type Tainted<'T> (context : TaintedContext, value : 'T) =
```
Comment (dsyme) [inline src/fsharp/tainted.fs:83]: No space before `:` please

**Review by dsyme — CHANGES_REQUESTED** [review]
I left a few comments.  It feels like too much is being made public.  Can we have baseline updates too please?

File: src/fsharp/tainted.fs (hunk 5 above)
Comment (DedSec256) [reply src/fsharp/tainted.fs:83]: Tainted is currently used in TaintedEquals methods in provided entities. If we transfer this logic, for example, inside Tainted, then we could make Tainted internal back

**Comment by auduchinok** [comment]
@DedSec256 Could you address the issues please?

**Comment by dsyme** [comment]
I think this can be closed now.  I just can't see us exposing this extension point from FCS - if we were to expose anything it would be a `ITypeProvider -> ITypeProvider` intercept that wraps a function around all type providers, allowing the FCS host to do whatever it likes with any `ITypeProvider`.  This doesn't need all these Provided* types to be exposed.

**Comment by DedSec256** [comment]
We have some ideas for this. I think we'll be back soon (:

**Pull request closed without merging**